src/System/Console/GetOpt/Generics.hs

```diff
@@ -118,17 +135,22 @@ mkEmptyArguments fields = case (sing :: Sing xs, fields) of
 -- * showing help?
 
 helpWrapper :: (All Option xs) =>
-  [String] -> NP FieldInfo xs -> a -> Either (IO ()) a
-helpWrapper args fields a =
+  String -> [Hint] -> [String] -> NP FieldInfo xs -> Either [String] a -> Result a
+helpWrapper header hints args fields result =
     case getOpt Permute [helpOption] args of
-      ([], _, _) -> Right a
-      (() : _, _, _) -> Left $ do
-        progName <- getProgName
-        let header = progName
-        putStrLn (usageInfo header (mkOptDescrs fields))
+      ([], _, _) -> case result of
```

Comment:
Mostly it's `getOpt`s fault! I pretty much have to read the documentation for it to figure out what each field of the tuple means, and even then it's not entirely obvious that `NoArg ()` would return `[()]` rather than `[]` on a match.

But maybe having a local where-clause:
```
hasAnyKnownArg (x, _, _) = not $ null x
```
Would help make clear what the first element of the tuple means?

Also, I'm correct in thinking that `"show help and exit"` isn't actually doing anything in this case, no?